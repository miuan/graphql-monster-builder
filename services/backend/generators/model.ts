import {
  SchemaModel,
  SchemaModelRelationType,
  StructureBackend,
  SchemaModelMember,
} from '../../common/types';

import {
  writeToFile,
} from '../../common/files';

const defaultMembers = [
  'createdAt',
  'updatedAt',
  'id',
];

import logger from '../../log'
import { BackendDirectory } from '../backendDirectory';
const log = logger.getLogger('model')

export const createMongoModel = (structure: StructureBackend, model : SchemaModel) => {
  const modelName = model.modelName;
  const lower = modelName.charAt(0).toLowerCase() + modelName.slice(1);
  const varName =  lower + 'Schema';
  let result = `import { Schema, Model, model } from 'mongoose'
  import { ${modelName}Model } from '../model-types'
  
const ${varName}: Schema = new Schema(
{
`;

  for (const member of model.members) {
    if (defaultMembers.indexOf(member.name) !== -1) {
      continue;
    }

    result += `\t\t${member.name}:`;
    
    let params = `{ type: ${transformTypeToMongoType(structure, member)}`;

    if (member.isRequired) {
      params += `, required: true`;
    }

    if (member.isUnique === true) {
      params += `, unique: true`;
    }

    if (member.default) {
      params += `, default: '${member.default}'`;
    }

    params += ' }';

    if (member.isArray) {
      result += ` [${params}],\n`;
    } else {
      result += ` ${params},\n`;
    }
    
  }

  if (modelName === 'User') {
    result += `__token: { type: Schema.Types.String, required: false},\n`
    result += `__refreshToken: { type: Schema.Types.String, required: false},\n`
    result += `__verifyToken: { type: Schema.Types.String, required: false},\n`
    result += `__password: { type: Schema.Types.String, required: true},\n`
    result += `__resetPasswordToken: { type: Schema.Types.String},\n`
    result += `__parent_access_token: { type: Schema.Types.String},\n`
  }

  result += 
`},
{
  timestamps: true,
  usePushEach: true,
  versionKey: false,
});

${varName}.pre('find', function() {
  (<any>this)._startTime = Date.now();
});

${varName}.post('find', function() {
  if ((<any>this)._startTime != null) {
    // console.log('Runtime in MS: ', Date.now() - (<any>this)._startTime);
  }
})

${varName}.pre('findOne', function() {
  (<any>this)._startTime = Date.now();
});

${varName}.post('findOne', function() {
  if ((<any>this)._startTime != null) {
    // console.log('Runtime in MS: ', Date.now() - (<any>this)._startTime);
  }
})

${varName}.pre('update', function() {
  (<any>this)._startTime = Date.now();
});

${varName}.post('update', function() {
  if ((<any>this)._startTime != null) {
    // console.log('Runtime in MS: ', Date.now() - (<any>this)._startTime);
  }
})
`

if(modelName === 'User'){
  // https://docs.mongodb.com/manual/core/index-partial/#examples
  result += `
  ${varName}.index(
    { __resetPasswordToken: 1 },
    { unique: true, partialFilterExpression: { __resetPasswordToken: { $exists: true } } }
  )

  ${varName}.index(
    { __verifyToken: 1 },
    { unique: true, partialFilterExpression: { __verifyToken: { $exists: true } } }
  )
 
`
}

// indexes
for (const member of model.members) {
  const isUnique = member.isUnique
  if(isUnique && isUnique[0]){
    result += `${varName}.index({ ${member.name}: 1`
    for(const c of isUnique as string[]){
      result += `,${c}: 1`
    }
    result += `}, { unique: true });`
  }
}

result +=`
export const ${lower}Model: Model<${modelName}Model> = model<${modelName}Model>('${structure.id}_${modelName}', ${varName});`;
  return result;
};

export const transformTypeToMongoType = (structure: StructureBackend, member : SchemaModelMember) => {
  
  if (member.relation) {
    let result = 'Schema.Types.ObjectId, ';
    result += `ref: '${structure.id}_${member.relation.relatedModel.modelName}', index: true`;
    return result;
  } 
  
  if (member.type === 'DateTime') {
    return 'Schema.Types.Date';
  } else if (member.type === 'Int') {
    return `Schema.Types.Number`;
  } else {
    return `Schema.Types.${member.modelName}`;
  }
};

export const generateMongoModelToFile = (backendDirectory: BackendDirectory, model: SchemaModel) => {
  const str = createMongoModel(backendDirectory.structure, model);

  backendDirectory.modelsWrite(`${model.modelName}`, str);
};


export const generateModels = (backendDirectory: BackendDirectory, models: SchemaModel[]) => {
  log.trace('generateModels')
  for (const model of models) {
    log.info(`Generate model: ${model.modelName}`)
    generateMongoModelToFile(backendDirectory, model);
  }
};
