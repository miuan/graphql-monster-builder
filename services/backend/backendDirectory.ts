import path = require("path")
import * as files from "../common/files"
import { Structure } from "../common/types"
  
export const BACKEND_STRUCTURE = {
    schema: { 
        dir: `ID/`,
        modules: ['schema.protectql'],
    },
    gen: { 
        dir: `ID/gen/`,
        modules: [],
    },
    models: { 
        dir:`ID/gen/models/`, 
        modules: [],
    },
    resolvers: { 
        dir:`ID/gen/resolvers/`, 
        modules: [],
    },
    services: { 
        dir:'ID/gen/services/', 
        modules: [],
    },
};
  
export interface BackendStructureOperatorWrite {
    write: (file: string, data: string) => boolean
    genWrite: (file: string, data: string) => boolean
    modelsWrite: (file: string, data: string) => boolean
    resolversWrite: (file: string, data: string) => boolean
    servicesWrite: (file: string, data: string) => boolean
}

export class BackendDirectory implements BackendStructureOperatorWrite {
    structure: Structure
    baseDir: string
    projectName: string

    constructor() {
        this.structure = JSON.parse(JSON.stringify(BACKEND_STRUCTURE)) as Structure
    }

    /**
     * 
     * @param projectName project name
     * @param baseDir base dir
     */
    init(projectName: string, baseDir: string = './graphql/generated/'){
        this.projectName = projectName
        this.baseDir = baseDir

        const outDir = projectName
        this.structure.id = projectName.replace('../', '');
      
        for (const str in this.structure) {
          if (str === 'id') {
            continue;
          }
      
          const item = this.structure[str];
          item.dir = path.join(baseDir, item.dir.replace('ID', outDir));
        }
    }

    /**
     * create or clean directory for copy files
     */
    prepareDirectory(){
        if(!this.baseDir || !this.projectName) {
            throw new Error('this.projectName or this.baseDir are empty, call init() fist')
        }

        files.removeDirs(path.join(this.baseDir, this.projectName, 'gen'));
        files.createDirs(this.structure)
    }
    
    public write(file: string, data: string){
        return files.writeToFile(this.structure.schema, file, data)
    }

    public genWrite(file: string, data: string){
        return files.writeToFile(this.structure.gen, file, data)
    }

    modelsWrite(file: string, data: string){
        return files.writeToFile(this.structure.models, file, data)
    }

    resolversWrite(file: string, data: string){
        return files.writeToFile(this.structure.resolvers, file, data)
    }

    servicesWrite(file: string, data: string){
        return files.writeToFile(this.structure.services, file, data)
    }
  
  }