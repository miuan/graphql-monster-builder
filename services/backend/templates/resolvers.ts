/**
 * Generated by GraphQL Monster Builder (GMB) version: __VERSION__
 *     Do NOT edit! For more info see: https://github.com/miuan/graphql-monster-builder
 */
import * as extras from '../extras'
import { UnauthorizedError, RequestError } from '../api-utils'

export const _MODEL_LOWER_NAME_All = (entry, protections) => {
    return async (root, data, ctx) => {
        _PROTECT_ALL_

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['before_MODEL_NAME_All']) {
            await entry.hooks.resolvers['before_MODEL_UPPER_NAME_All'](entry, { root, data, ctx })
        }

        let models = await entry.services['_MODEL_LOWER_NAME_'].all(data, ctx.userId)
        models = models.map((m) => {
            m.id = m._id
            return m
        })

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['after_MODEL_UPPER_NAME_All']) {
            models = await entry.hooks.resolvers['after_MODEL_UPPER_NAME_All'](entry, { models, root, data, ctx })
        }
        return models
    }
}

export const _MODEL_LOWER_NAME_Count = (entry, protections) => {
    return async (root, data, ctx) => {
        _PROTECT_ALL_

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['before_MODEL_NAME_Count']) {
            await entry.hooks.resolvers['before_MODEL_UPPER_NAME_Count'](entry, { root, data, ctx })
        }

        let models = await entry.services['_MODEL_LOWER_NAME_'].count(data, ctx.userId)

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['after_MODEL_UPPER_NAME_Count']) {
            models = await entry.hooks.resolvers['after_MODEL_UPPER_NAME_Count'](entry, { models, root, data, ctx })
        }
        return models
    }
}

export const _MODEL_LOWER_NAME_One = (entry, protections) => {
    return async (root, data, ctx) => {
        _PROTECT_ONE_
        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['before_MODEL_UPPER_NAME_One']) {
            await entry.hooks.resolvers['before_MODEL_UPPER_NAME_One'](entry, { root, data, ctx })
        }
        let model = await entry.services['_MODEL_LOWER_NAME_'].one(data.id)

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['after_MODEL_UPPER_NAME_One']) {
            model = await entry.hooks.resolvers['after_MODEL_UPPER_NAME_One'](entry, { id: data.id, model, entry, root, data, ctx })
        }

        return model
    }
}

export const _MODEL_LOWER_NAME_Create = (entry, protections) => {
    return async (root, data, ctx) => {
        _PROTECT_CREATE_
        const presentProtectedFields = protections.checkDataContainProtectedFields(data)
        if (presentProtectedFields && presentProtectedFields.length) {
            ctx.throw(403, {
                type: 'ReachProtectedFields',
                message: 'Trying to update protected fields, which they are just read only',
                presentProtectedFields,
            })
        }

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['before_MODEL_UPPER_NAME_Create']) {
            data = await entry.hooks.resolvers['before_MODEL_UPPER_NAME_Create'](entry, { root, data, ctx })
        }

        const userId = ctx.state?.user?.id
        _AUTOMATIC_USER_FROM_CTX_
        let createdModel = await entry.services['_MODEL_LOWER_NAME_'].create(data, userId)

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['after_MODEL_UPPER_NAME_Create']) {
            createdModel = await entry.hooks.resolvers['after_MODEL_UPPER_NAME_Create'](entry, { root, data, ctx, createdModel })
        }

        return createdModel
    }
}

export const _MODEL_LOWER_NAME_Update = (entry, protections) => {
    return async (root, data, ctx) => {
        _PROTECT_UPDATE_

        // any fields start with double underscore are protected, example __port, __readolny, ...
        const presentProtectedFields = protections.checkDataContainProtectedFields(data)
        if (presentProtectedFields && presentProtectedFields.length) {
            ctx.throw(403, {
                type: 'ReachProtectedFields',
                message: 'Trying to update protected fields, which they are just read only',
                presentProtectedFields,
            })
        }

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['before_MODEL_UPPER_NAME_Update']) {
            data = await entry.hooks.resolvers['before_MODEL_UPPER_NAME_Update'](entry, { root, data, ctx })
        }

        let updatedModel = await entry.services['_MODEL_LOWER_NAME_'].update(data, null, ctx.userId)

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['after_MODEL_UPPER_NAME_Update']) {
            updatedModel = await entry.hooks.resolvers['after_MODEL_UPPER_NAME_Update'](entry, { root, data, ctx, updatedModel, id: data.id })
        }

        return updatedModel
    }
}

export const _MODEL_LOWER_NAME_Remove = (entry, protections) => {
    return async (root, data, ctx) => {
        _PROTECT_REMOVE_
        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['before_MODEL_UPPER_NAME_Remove']) {
            await entry.hooks.resolvers['before_MODEL_UPPER_NAME_Remove'](entry, { root, data, ctx })
        }

        let removedModel = await entry.services['_MODEL_LOWER_NAME_'].remove(data.id, ctx.userId)

        if (entry.hooks && entry.hooks.resolvers && entry.hooks.resolvers['after_MODEL_UPPER_NAME_Remove']) {
            removedModel = await entry.hooks.resolvers['after_MODEL_UPPER_NAME_Remove'](entry, { removedModel, root, data, ctx })
        }

        return removedModel
    }
}

_RESOLVERS_ADD_REMOVE_

export const generate_MODEL_NAME_Resolver = (entry) => {
    const protections = extras.generateProtections(entry, '_MODEL_LOWER_NAME_')
    return {
        all: _MODEL_LOWER_NAME_All(entry, protections),
        count: _MODEL_LOWER_NAME_Count(entry, protections),
        one: _MODEL_LOWER_NAME_One(entry, protections),
        create: _MODEL_LOWER_NAME_Create(entry, protections),
        update: _MODEL_LOWER_NAME_Update(entry, protections),
        remove: _MODEL_LOWER_NAME_Remove(entry, protections),
        _RESOLVERS_ADD_REMOVE_CONNECT_,
    }
}
