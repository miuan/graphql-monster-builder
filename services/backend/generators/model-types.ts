import {
  SchemaModel,
  SchemaModelMember,
} from '../../common/types';


const defaultMembers = [
  'createdAt',
  'updatedAt',
  'id',
];

import logger from '../../log'
import { BackendDirectory } from '../backendDirectory';
const log = logger.getLogger('types')

export const createModelTypeFromModel = (model : SchemaModel) => {
  const {modelName} = model
  let modelMembers = ''

  for (const member of model.members) {
    modelMembers += `\n\t\t${member.name}: ${transformMemberTypeToTypescriptType(member)}`
  }

  if (modelName === 'User') {
    modelMembers += `\n\t\t__token?: string,\n`
    modelMembers += `\n\t\t__refreshToken?: string,\n`
    modelMembers += `\n\t\t__verifyToken?: string,\n`
    modelMembers += `\n\t\t__password: string,\n`
    modelMembers += `\n\t\t__resetPasswordToken: string,\n`
    modelMembers += `\n\t\t__parent_access_token: string,\n`
  }

  return `
    // type for model: \`${modelName}\`
    export interface ${modelName}Model extends Document {${modelMembers}
    }
  `
}
  
export const transformMemberTypeToTypescriptType = (member : SchemaModelMember) => {
  let ttype

  if (member.relation) {
     ttype = `${member.relation.relatedModel.modelName}Model`;
  } else if (member.type === 'DateTime') {
    ttype = 'Date';
  } else if (member.type === 'Int' || member.type === 'Float') {
    ttype =`number`;
  } else if (member.type === 'ID' || member.type === 'String') {
    ttype =`string`;
  } else {
    ttype = `${member.type}`;
  }

  if(member.isArray){
    ttype+='[]'
  }

  return ttype
};

export const generateModelTypes = (backendDirectory: BackendDirectory, models: SchemaModel[]) => {
  log.trace('generateTypescriptTypes')
  let typesAll = `import { Document } from 'mongoose'\n`
  for (const model of models) {
    typesAll += createModelTypeFromModel(model);
  }
  log.info(`Generate types for all models: types.ts`)
  backendDirectory.genWrite(`model-types.ts`, typesAll);
};
