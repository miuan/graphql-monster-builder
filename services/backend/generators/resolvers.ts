import { templateFileToText, templateToText } from '../../common/files'
import { SchemaModel, SchemaModelMember, SchemaModelProtectionParam, SchemaModelProtectionType, SchemaModelRelationType, SYSTEM_MODELS } from '../../common/types'
import { firstToLower, getOnlyOneRelatedMember } from '../../common/utils'
import logger from '../../log'
import { BackendDirectory } from '../backendDirectory'

const log = logger.getLogger('resolvers')

const defaultMembers = ['createdAt', 'updatedAt', 'id']

const memberCreateAndRemoveLinks = (model: SchemaModel, member: SchemaModelMember) => {
    const modelName = model.modelName
    const relation = member.relation
    const linkNames = relation.linkNames

    const ret = {
        result: '',
        connect: '',
    }

    const lower = firstToLower(modelName)
    const funcLinkName = linkNames.linkName
    const funcUnlinkName = linkNames.unlinkName

    let protection
    let unlinkProtection = ''
    if (relation.name === '_RoleToUser') {
        const conditions = constructConditionsFromProtection([{ type: SchemaModelProtectionType.ROLE, roles: ['admin'] }])
        protection = compileConditionsToIfStatement(conditions)
        unlinkProtection = `
            if(ctx.state?.user?.id == data.userId){
                if(data.userRoleName === 'admin') {
                    throw new Error('Unlinking yourself from admin')
                } else {
                    const role = await entry.models['userRole'].findById(data.userRoleId, 'name').lean()
                    if(role.name === 'admin') {
                        throw new Error('Unlinking yourself from admin')
                    }
                }
            }
        `
    } else {
        const conditions1 = constructConditionsFromProtection(model.protection.update, linkNames.param1, modelName)
        const conditions2 = constructConditionsFromProtection(relation.relatedModel.protection.update, linkNames.param2, relation.relatedModel.modelName)
        const conditions = [...conditions1, ...conditions2.filter((c2) => !conditions1.includes(c2))]
        protection = compileConditionsToIfStatement(conditions)
    }

    const params = linkNames.param3 ? `data.${linkNames.param1}, data.${linkNames.param2}, data.${linkNames.param3}` : `data.${linkNames.param1}, data.${linkNames.param2}`

    ret.result = templateFileToText('resolvers-add-remove.ts', {
        _LOWER_NAME_: lower,
        _FUNC_LINK_NAME_: funcLinkName,
        _FUNC_UNLINK_NAME_: funcUnlinkName,
        _PARAMS_: params,
        _UNLINK_PROTECTION_: unlinkProtection,
        _PROTECTION_: protection,
    })

    ret.connect += `${funcLinkName} : ${funcLinkName}(entry, protections),\n${funcUnlinkName} : ${funcUnlinkName}(entry, protections),`

    return ret
}

const modelCreateAddRemoveLinks = (model: SchemaModel) => {
    const membersWithRelations = model.members.filter((member) => member.relation?.linkNames)
    const ret = {
        result: '',
        connect: '',
    }

    for (const member of membersWithRelations) {
        const { result, connect } = memberCreateAndRemoveLinks(model, member)

        ret.result += result
        ret.connect += connect
    }
    return ret
}

export const createResolver = (model: SchemaModel) => {
    const modelName = model.modelName
    const lower = modelName.charAt(0).toLowerCase() + modelName.slice(1)
    const varName = lower + 'Service'

    const protectionAll = compileConditionsToIfStatement(constructConditionsFromProtection(model.protection.all, modelName), modelName)
    const protectionOne = compileConditionsToIfStatement(constructConditionsFromProtection(model.protection.one, modelName), modelName)
    const protectionCreate = compileConditionsToIfStatement(constructConditionsFromProtection(model.protection.create, modelName), modelName)
    const protectionUpdate = compileConditionsToIfStatement(constructConditionsFromProtection(model.protection.update, modelName), modelName)
    const protectionRemove = compileConditionsToIfStatement(constructConditionsFromProtection(model.protection.remove, modelName), modelName)

    const { result: _RESOLVERS_ADD_REMOVE_, connect: _RESOLVERS_ADD_REMOVE_CONNECT_ } = modelCreateAddRemoveLinks(model)

    const automaticUserFromCtx = !SYSTEM_MODELS.includes(modelName)
        ? `if(!data.userId && userId){
        data.userId = userId
      }`
        : ''

    const file = templateFileToText('resolvers.ts', {
        _RESOLVERS_ADD_REMOVE_CONNECT_,
        _RESOLVERS_ADD_REMOVE_,
    })

    const modelUpperName = modelName[0].toUpperCase() + modelName.substr(1)

    let result = templateToText(file, {
        _MODEL_NAME_: modelName,
        _MODEL_UPPER_NAME_: modelUpperName,
        _MODEL_LOWER_NAME_: lower,
        _PROTECT_ALL_: protectionAll,
        _PROTECT_ONE_: protectionOne,
        _PROTECT_CREATE_: protectionCreate,
        _PROTECT_UPDATE_: protectionUpdate,
        _PROTECT_REMOVE_: protectionRemove,
        _AUTOMATIC_USER_FROM_CTX_: automaticUserFromCtx,
    })

    result += ''
    return result
}

export const compileConditionsToIfStatement = (conditions: string[], modelName: string = undefined) => {
    const condition = conditions.join('&&')
    let result = `if(${condition}){
        throw new Error('Unauthorized')
      }`

    if (modelName && modelName == 'User') {
        result += `
    if((data.roles || data.rolesIds) && !await protections.role(ctx, ['admin'])){
        throw new Error('Unauthorized user:roles operation')
    }
  `
    }

    return result
}

export const constructConditionsFromProtection = (protections: SchemaModelProtectionParam[], modelName: string = undefined, objectIdField = undefined) => {
    return protections.map((protectionParam) => `!(${generateProtectionFromParam(protectionParam, modelName, objectIdField)})`)
}

export const generateProtectionFromParam = (protection: SchemaModelProtectionParam, modelName: string = undefined, objectIdField = undefined) => {
    let result = ''

    if (protection.type === SchemaModelProtectionType.PUBLIC) {
        result += `await protections.public()`
    } else if (protection.type === SchemaModelProtectionType.USER) {
        result += `await protections.user(ctx)`
    } else if (protection.type === SchemaModelProtectionType.OWNER) {
        // in model user is owner identified by ID
        const param = modelName && modelName === 'User' ? '_id' : 'user'
        result += objectIdField ? `await protections.owner(ctx, data, '${param}', '${objectIdField}')` : `await protections.owner(ctx, data, '${param}')`
    } else if (protection.type === SchemaModelProtectionType.ROLE) {
        const roles = `'` + protection.roles.join(`','`) + `'`
        result += `await protections.role(ctx, [${roles}])`
    } else if (protection.type === SchemaModelProtectionType.FILTER) {
        const roles = protection.roles.length > 0 ? `['` + protection.roles.join(`','`) + `']` : 'null'
        let filters = ''
        for (const filter of protection.filter) {
            filters = `,{name: '${filter.name}', value: '${filter.value}'}`
        }

        result += `await protections.filter(ctx, data, [${filters.substr(1)}], ${roles})`
    }
    return result
}

export const generateResolverToFile = (backendDirectory: BackendDirectory, model: SchemaModel) => {
    const str = createResolver(model)

    backendDirectory.resolversWrite(`${model.modelName}`, str)
}

export const generateResolvers = (backendDirectory: BackendDirectory, models: SchemaModel[]) => {
    log.trace('generateResolvers')
    for (const model of models) {
        log.info(`Generate resolver for model: ${model.modelName}`)
        generateResolverToFile(backendDirectory, model)
    }
}
