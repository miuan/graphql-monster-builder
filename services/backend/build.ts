import { templateFileToText} from '../common/files';
import { getModelsFromSchema } from '../parser/scan';
import { generateModels } from './generators/model';
import { generateServices } from './generators/service';
import { generateResolvers } from './generators/resolvers';
import { generateDataloaders } from './generators/dataloaders';
import { generateEntry } from './generators/entry';
import * as fs from 'fs';

import { generateSchema } from './generators/schema';
import { generateModelTypes } from './generators/model-types';
import { BackendDirectory } from './backendDirectory';
import * as _ from 'lodash'
import { generateIntegrationTests } from './generators/integration';

export const exportAs = async (name, from, base='.', config={}) => {
  
  const importedSchema = fs.readFileSync(from, { encoding: 'utf8' });
  return exportAsFromString(name, importedSchema, base, config)
}

export const exportAsFromString = async (name, importedSchema, outDir='.', config={}) => {
  const models = await getModelsFromSchema(importedSchema);
 
  const backendDirectory = new BackendDirectory()
  backendDirectory.init(name, outDir)
  // TODO: in some times 'services/' is not used,
  //       so not create directory what is not necessary
  backendDirectory.prepareDirectory()

  generateSchema(backendDirectory, models);
  generateModels(backendDirectory, models);
  generateServices(backendDirectory, models);
  generateResolvers(backendDirectory, models);
  generateDataloaders(backendDirectory, models);
  generateModelTypes(backendDirectory, models);
  generateIntegrationTests(backendDirectory, models);

  generateEntry(backendDirectory, models);

  backendDirectory.writeWithConfig('server', config)
  backendDirectory.writeWithConfig('sendMail', config, 'services')
  backendDirectory.writeWithConfig('passport', config, 'services')
  backendDirectory.writeWithConfig('extras', config, 'gen');

  backendDirectory.genWrite( `schemaLoad`, templateFileToText(`schemaLoad.ts`, null));

  backendDirectory.genWrite( `services/db`, templateFileToText(`db.ts`, null));
  backendDirectory.genWrite( `source.schema`, importedSchema);
  backendDirectory.write( `package.json`, templateFileToText(`package.json`, null));
  
}
