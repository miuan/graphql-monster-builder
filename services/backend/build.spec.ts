import { SchemaModel, SchemaModelRelation, SchemaModelType } from "../common/types"
import { BackendDirectory } from "./backendDirectory"
import { generateAll } from "./build"
import {generateMongoModelToFile} from './generators/model'
import {createModelTypeFromModel} from './generators/model-types'
import { generateSchema } from './generators/schema'
import { generateResolverToFile } from './generators/resolvers'
import { generateServiceToFile } from './generators/service'
import { SchemaType } from "mongoose"

jest.mock('./generators/model')
jest.mock('./generators/model-types')
jest.mock('./generators/schema')
jest.mock('./generators/resolvers')
jest.mock('./generators/service')

  describe('generate:build', ()=>{
       
    it('should call 2x createTypescriptType and 1x writeToFile inside generateTypescriptTypes', ()=>{
        const backendDirectory = new BackendDirectory()
        backendDirectory.genWrite = jest.fn()
        backendDirectory.modelsWrite = jest.fn()
        backendDirectory.servicesWrite = jest.fn()
        backendDirectory.resolversWrite = jest.fn()

        // (modelGenerator as any).generateMongoModelToFile = jest.fn()

        const modelTodo = {
          modelName: 'Todo',
          type: SchemaModelType.MODEL,
          members:[
            {
              name: 'email',
              type: 'String'
            },
            {
              name: 'virtual',
              type: 'String',
              isVirtual: true
            }
          ]
      } as SchemaModel

      const modelUser = {
        modelName: 'User',
        type: SchemaModelType.ENTITY,
        members: [{
            name: 'email',
            type: 'String'
        },{
            name: 'password',
            type: 'Boolean',
            isVirtual: true
        },{
            name: 'todos',
            type: 'Todo',
            isArray: true,
            relation:{
                relatedModel:{modelName: 'Todo'}
            } as SchemaModelRelation
        }]
    } as SchemaModel


        generateAll(backendDirectory, [modelTodo, modelUser], {})

        expect(generateSchema).toBeCalledTimes(1)

        expect(generateMongoModelToFile).toBeCalledTimes(2)
        expect((generateMongoModelToFile as any).mock.calls[0]).toEqual([backendDirectory, modelTodo, expect.not.arrayContaining([expect.objectContaining({isVirtual:true})])])
        expect((generateMongoModelToFile as any).mock.calls[1]).toEqual([backendDirectory, modelUser, expect.not.arrayContaining([expect.objectContaining({isVirtual:true})])])
        
        expect(createModelTypeFromModel).toBeCalledTimes(2)
        expect((createModelTypeFromModel as any).mock.calls).toEqual(expect.arrayContaining([
          [modelTodo, expect.not.arrayContaining([expect.objectContaining({isVirtual:true})])],
          [modelUser, expect.not.arrayContaining([expect.objectContaining({isVirtual:true})])]
        ]))
        
        
        expect(generateServiceToFile).toBeCalledTimes(2)
        expect((generateServiceToFile as any).mock.calls).toEqual(expect.arrayContaining([
          [backendDirectory, modelTodo],
          [backendDirectory, modelUser]
        ]))

        expect(generateResolverToFile).toBeCalledTimes(1)
        expect((generateResolverToFile as any).mock.calls).toEqual(expect.arrayContaining([
          [backendDirectory, expect.not.objectContaining({type: SchemaModelType.ENTITY})]
        ]))
        
        const modelTypesCall = (backendDirectory.genWrite as any).mock.calls.find((calls=>calls[0]=='model-types.ts'))
        expect(modelTypesCall).toBeDefined()

        const dataloadersCall = (backendDirectory.genWrite as any).mock.calls.find((calls=>calls[0]=='dataloaders'))
        expect(dataloadersCall).toBeDefined()

        const entryCall = (backendDirectory.genWrite as any).mock.calls.find((calls=>calls[0]=='entry'))
        expect(entryCall).toBeDefined()

    })

})

