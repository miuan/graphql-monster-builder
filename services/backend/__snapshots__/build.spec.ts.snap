// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema Export inputs student have two imputs 1`] = `
"input TodoItemimagesImageTodoItem {
url: String!
}
input ImageTodoItemtodoTodoItem {
completed: Boolean!, title: String!, userId: ID
}
input ClassstudentsStudent {
userId: ID, fullname: String!, age: Int
}
input StudentclassesClass {
name: String!
}
input UsertodoItemsTodoItem {
completed: Boolean!, title: String!, imagesIds: [ID!], images: [TodoItemimagesImageTodoItem!]
}
input UserrolesUserRole {
name: String!
}

"
`;

exports[`Schema Export mutations todo item have payloads for Add and Remove 1`] = `
"# on model TodoItem - user
type AddToTodoItemOnUserPayload {
  todoItemsTodoItem: TodoItem
  userUser: User
}

type RemoveFromTodoItemOnUserPayload {
  todoItemsTodoItem: TodoItem
  userUser: User
}
# on model TodoItem - images
type AddToImageTodoItemOnTodoItemPayload {
  todoTodoItem: TodoItem
  imagesImageTodoItem: ImageTodoItem
}

type RemoveFromImageTodoItemOnTodoItemPayload {
  todoTodoItem: TodoItem
  imagesImageTodoItem: ImageTodoItem
}
"
`;
